### PR TITLE
♻️ ooil: enhancements on PR#3428

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

``service-integration version``: 1.0.1 → 1.0.2

-  Changes in the design on PR #3428:
   -  ♻️ Merges ``IntegrationContext`` into already existing ``AppSettings``: central ``BaseSettings`` class for all settings init at startup!
   - 🐛 ``compose``:  ``docker_compose_overwrite_cfg`` is required but NOT its ``..._path`` counterpart
   - ♻️``DockerComposeOverwriteCfg`` two different factory functions for two different things: ``create_default`` and ``from_yaml``.
- ✨ adds new regex for semantic versions and removes ``basic_types``
- [x] Evaluate  https://github.com/ITISFoundation/cookiecutter-osparc-service:
   - 🐛 Fixes in tests and CLI based on usage https://github.com/ITISFoundation/cookiecutter-osparc-service/pull/40. The goal is to merge that PR and make this library compatible with the old cookie-cutter for services.
- [x] Evaluate https://github.com/ITISFoundation/ci-service-integration-library. 
   - only uses ``ooil compose``. Tested in ``tests/test_usecase_jupytermath.py``

## Related issue/s

- Follows up from #3428
- Implementing #3418 

## How to test

- Install and run tests
```cmd
cd packages/integration-library
make install-dev
make tests
```
- Install and use CLI to create a ``.osparc`` configuration by reading labels from a docker-compose. For instance:
```cmd
$ ooil config -f tests/data/docker-compose-meta.yml 
Creating tests/data/.osparc from tests/data/docker-compose-meta.yml ...
Creating osparc config files from tests/data/docker-compose-meta.yml
Creating tests/data/.osparc/dy-static-file-server/metadata.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server/docker-compose.overwrite.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server/runtime.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar/metadata.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar/docker-compose.overwrite.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar/runtime.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar-compose-spec/metadata.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar-compose-spec/docker-compose.overwrite.yml ...DONE
Creating tests/data/.osparc/dy-static-file-server-dynamic-sidecar-compose-spec/runtime.yml ...DONE
osparc config files created
```
- An now the opposite. Just modify some fields in `` tests/data/.osparc/dy-static-file-server/metadata.yml `` and produce a new ``docker-compose``
```cmd
$ ooil compose --metadata tests/data/.osparc/dy-static-file-server/metadata.yml --to-spec-file new-docker-compose.yaml
No explicit config for OCI/label-schema found (optional), skipping OCI annotations.
```
and can now observer ``new-docker-compose.yaml``

## Checklist

- [x] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: packages/service-integration/src/service_integration/settings.py
Comment (GitHK) [inline packages/service-integration/src/service_integration/settings.py:20]: This will break all the CIs we have.

Previously we had the following:
```
    REGISTRY_NAME: str = Field(
        "", description="name of the registry to use for images, default is Docker Hub"
    )
```
Please leave it to empty string meaning Docker Hub.

File: packages/service-integration/src/service_integration/versioning.py
Comment (GitHK) [inline packages/service-integration/src/service_integration/versioning.py:69]: Why add a default here since it previously had none?

**Review by GitHK — COMMENTED** [review]
Thanks for the changes. I've added a few comments to consider.

Please note that `ooil` is packaged inside this library https://github.com/ITISFoundation/ci-service-integration-library

The `ci-service-integration-library` is used for the following purposes:
- validating services like https://github.com/ITISFoundation/jupyter-math (see usages below)
   -https://github.com/ITISFoundation/jupyter-math/blob/main/Makefile#L14
   - https://github.com/ITISFoundation/jupyter-math/blob/main/Makefile#L28
   - https://github.com/ITISFoundation/jupyter-math/blob/main/.github/workflows/check-image.yml
- in the internal CI for building and publishing such services like the above. https://git.speag.com/oSparc/docker-publisher-osparc-services/-/blob/master/.gitlab-ci.yml#L2

So before merging these services could you please make a release of `ci-service-integration-library` and see If it still capable of running against the above use cases?

File: packages/service-integration/src/service_integration/settings.py
Comment (pcrespov) [reply packages/service-integration/src/service_integration/settings.py:20]: True. My mistake :-)

But i notice now that this change you added introduces an inconsistency with the previous settings ... I will try to adapt  both.

File: packages/service-integration/src/service_integration/versioning.py
Comment (pcrespov) [reply packages/service-integration/src/service_integration/versioning.py:69]: you have a valid point.

**Comment by pcrespov** [comment]
> Thanks for the changes. I've added a few comments to consider.
> 
> Please note that `ooil` is packaged inside this library https://github.com/ITISFoundation/ci-service-integration-library
> 
> The `ci-service-integration-library` is used for the following purposes:
> 
> * validating services like https://github.com/ITISFoundation/jupyter-math (see usages below)
>   
>   * https://github.com/ITISFoundation/jupyter-math/blob/main/Makefile#L14
>   * https://github.com/ITISFoundation/jupyter-math/blob/main/Makefile#L28
>   * https://github.com/ITISFoundation/jupyter-math/blob/main/.github/workflows/check-image.yml
> * in the internal CI for building and publishing such services like the above. https://git.speag.com/oSparc/docker-publisher-osparc-services/-/blob/master/.gitlab-ci.yml#L2
> 
> So before merging these services could you please make a release of `ci-service-integration-library` and see If it still capable of running against the above use cases?
> 
> I can give you a hand with all this to make sure everything is working as expected and we avoid breaking existing pipelines.


Yes, I am doing the same with  https://github.com/ITISFoundation/cookiecutter-osparc-service

File: packages/service-integration/tests/test__usecase_jupytermath.py
Comment (GitHK) [inline packages/service-integration/tests/test__usecase_jupytermath.py:87]: Very nice!

File: packages/service-integration/tests/test__usecase_jupytermath.py
Comment (GitHK) [inline packages/service-integration/tests/test__usecase_jupytermath.py:110]: You could add an additional layer of testing to make sure there are not regressions.

Run [make compose-spec](https://github.com/ITISFoundation/jupyter-math/blob/ad51f531548e88afa3ffe3b702a89d159ad8be7f/Makefile#L25) in the cloned repository. Then compare the generated `docker-compose.yaml` file with the locally generated one.

Comment (pcrespov) [reply packages/service-integration/tests/test__usecase_jupytermath.py:110]: OK, i get your point but I have done some modifications to your proposal:

Pulling ``Unable to find image 'itisfoundation/ci-service-integration-library:v1.0.1-dev-25' locally`` takes a lot of time for this type of tests. I have instead pinned the repo commit and taken the resulting ``docker-compose``and put it in ``tests/data/docker-compose_jupyter-math_v1-0-1-dev-25.yml``

In addition, take into account that this test is to ensure that ``ooil`` HEAD works as before this date!  Nonetheless, newer versions of ``ooil`` might produce in time different ``docker-compose`` outputs to those generated by an older version. This means that the comparison is far more subtle than just diff-ing two docker-compose ! We have to check instead that those are "equivalent" (since, as I mentioned, they well be produced by different versions of ``ooil``)

IMO, the reasons above suggest that the test, as you suggested, is more suited on the client side instead of here.

Here I will add a variation of your suggestion that I think reasonably cover our purpose.

Comment (GitHK) [reply packages/service-integration/tests/test__usecase_jupytermath.py:159]: Perfect, this will do the job as well. Thanks for taking care of it!
This should give us some guarantees about the process!

**Review by GitHK — APPROVED** [review]
👍 Very nice overall. I'm glad we have some extra checks in place.

**Review by sanderegg — APPROVED** [review]
pair reviewed

**Pull request merged**
